src/bit_machine/tracker.rs

```diff
@@ -0,0 +1,205 @@
+    ///     the write frame after the node has executed. See [`NodeOutput`] for more information.
+    fn visit_node(&mut self, _node: &RedeemNode<J>, _input: super::FrameIter, _output: NodeOutput) {
+    }
+}
+
+pub trait PruneTracker<J: Jet>: ExecTracker<J> {
+    /// Returns true if the left branch of the of the `Case` node with the IHR `ihr` was taken.
+    fn contains_left(&self, ihr: Ihr) -> bool;
+
+    /// Returns true if the left branch of the of the `Case` node with the IHR `ihr` was taken.
+    fn contains_right(&self, ihr: Ihr) -> bool;
+}
+
+/// Tracker of executed left and right branches for each case node.
+#[derive(Clone, Debug, Default)]
+pub struct SetTracker {
+    left: HashSet<Ihr>,
+    right: HashSet<Ihr>,
+}
+
+impl<J: Jet> ExecTracker<J> for SetTracker {
+    fn visit_node<'d>(
+        &mut self,
+        node: &RedeemNode<J>,
+        mut input: super::FrameIter,
+        _output: NodeOutput,
+    ) {
+        match (node.inner(), input.next()) {
+            (Inner::AssertL(..) | Inner::Case(..), Some(false)) => {
+                self.left.insert(node.ihr());
+            }
+            (Inner::AssertR(..) | Inner::Case(..), Some(true)) => {
+                self.right.insert(node.ihr());
+            }
+            _ => {}
+        }
+    }
+}
+
+impl<J: Jet> PruneTracker<J> for SetTracker {
+    fn contains_left(&self, ihr: Ihr) -> bool {
+        self.left.contains(&ihr)
+    }
+
+    fn contains_right(&self, ihr: Ihr) -> bool {
+        self.right.contains(&ihr)
+    }
+}
+
+/// Tracker that does not do anything (noop).
+#[derive(Copy, Clone, Debug)]
+pub struct NoTracker;
+
+impl<J: Jet> ExecTracker<J> for NoTracker {
+    fn visit_node<'d>(
+        &mut self,
+        node: &RedeemNode<J>,
+        mut input: super::FrameIter,
+        output: NodeOutput,
+    ) {
+        if cfg!(test) {
+            // In unit tests, attempt to decode values from the frames, confirming that
+            // decoding works.
+            let _input_val = Value::from_padded_bits(&mut input, &node.arrow().source);
+            if let NodeOutput::Success(mut output) = output {
+                let _output_val = Value::from_padded_bits(&mut output, &node.arrow().source);
+            }
+        }
+    }
+}
+
+/// Tracker that just outputs all its activity to stderr.
+#[derive(Clone, Debug, Default)]
+pub struct StderrTracker {
+    exec_count: usize,
+    inner: SetTracker,
+}
+
+impl StderrTracker {
+    /// Constructs a new empty [`StderrTracker`], ready for use.
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+impl<J: Jet> ExecTracker<J> for StderrTracker {
+    fn visit_node(
+        &mut self,
+        node: &RedeemNode<J>,
+        mut input: super::FrameIter,
+        output: NodeOutput,
+    ) {
+        let input_val = Value::from_padded_bits(&mut input, &node.arrow().source)
+            .expect("input from bit machine will always be well-formed");
```

Comment:
```suggestion
    ) {
    	self.inner.visit_node(node, input.clone(), output.clone());
        
        let input_val = Value::from_padded_bits(&mut input, &node.arrow().source)
            .expect("input from bit machine will always be well-formed");
```

Is it ok, that we do not have `self.inner.visit_node(node, input.clone(), output.clone());` here?

Reply:
oo, no, it's not! Good catch.